packages/block-library/src/navigation/edit/edit-overlay-button.js

```diff
@@ -0,0 +1,172 @@
+/**
+ * WordPress dependencies
+ */
+import { __ } from '@wordpress/i18n';
+import {
+	Button,
+	MenuGroup,
+	MenuItem,
+	MenuItemsChoice,
+	DropdownMenu,
+} from '@wordpress/components';
+import { useSelect, useDispatch } from '@wordpress/data';
+import { privateApis as routerPrivateApis } from '@wordpress/router';
+import { store as coreStore } from '@wordpress/core-data';
+import { parse, serialize } from '@wordpress/blocks';
+import { moreVertical } from '@wordpress/icons';
+/**
+ * Internal dependencies
+ */
+import { unlock } from '../../lock-unlock';
+import useGoToOverlayEditor from './use-go-to-overlay-editor';
+import useOverlay from './use-overlay';
+
+const { useHistory } = unlock( routerPrivateApis );
+
+export default function EditOverlayButton( {
+	navRef,
+	attributes,
+	setAttributes,
+} ) {
+	const currentOverlayId = attributes?.overlayId;
+
+	// Get any custom overlay attached to this block,
+	// falling back to the one provided by the Theme.
+	const overlay = useOverlay( currentOverlayId );
+
+	const { coreOverlay, allOverlays } = useSelect( ( select ) => {
+		return {
+			// Get the default template part that core provides.
+			coreOverlay: select( coreStore ).getEntityRecord(
+				'postType',
+				'wp_template_part',
+				`core//navigation-overlay`
+			),
+			// Get all the overlays.
+			allOverlays: select( coreStore ).getEntityRecords(
+				'postType',
+				'wp_template_part',
+				{
+					area: 'navigation-overlay',
+				}
+			),
+		};
+	}, [] );
+
+	const { saveEntityRecord } = useDispatch( coreStore );
+
+	const history = useHistory();
+
+	const goToOverlayEditor = useGoToOverlayEditor();
+
+	async function handleEditOverlay( event ) {
+		event.preventDefault();
+
+		// There may already be an overlay with the slug `navigation-overlay`.
+		// This might be a user created one, or one provided by the theme.
+		// If so, then go directly to the editor for that overlay template part.
+		if ( overlay ) {
+			goToOverlayEditor( overlay.id, navRef );
+			return;
+		}
+
+		// If there is not overlay then create one using the base template part
+		// provided by Core.
+		// TODO: catch and handle errors.
+		const overlayBlocks = buildOverlayBlocks( coreOverlay.content.raw );
+
+		// The new overlay should use the current Theme's slug.
+		const newOverlay = await createOverlay( overlayBlocks );
+
```

Comment:
The overlay should be assigned to the block attributes here. Then the attribute can be used for backwards compatibility in index.php (if it has the attribute, do the new thing, else do the old thing)